diff --git a/clang/include/clang/Driver/Options.td b/clang/include/clang/Driver/Options.td
index 958d0d05a..99a525e1b 100644
--- a/clang/include/clang/Driver/Options.td
+++ b/clang/include/clang/Driver/Options.td
@@ -8588,11 +8588,6 @@ def objc_isystem : Separate<["-"], "objc-isystem">,
 def objcxx_isystem : Separate<["-"], "objcxx-isystem">,
   MetaVarName<"<directory>">,
   HelpText<"Add directory to the ObjC++ SYSTEM include search path">;
-def internal_iframework : Separate<["-"], "internal-iframework">,
-  MetaVarName<"<directory>">,
-  HelpText<"Add directory to the internal system framework search path; these "
-           "are assumed to not be user-provided and are used to model system "
-           "and standard frameworks' paths.">;
 def internal_isystem : Separate<["-"], "internal-isystem">,
   MetaVarName<"<directory>">,
   HelpText<"Add directory to the internal system include search path; these "
diff --git a/clang/include/clang/Driver/ToolChain.h b/clang/include/clang/Driver/ToolChain.h
index b8899e781..5fe9a5dcb 100644
--- a/clang/include/clang/Driver/ToolChain.h
+++ b/clang/include/clang/Driver/ToolChain.h
@@ -225,9 +225,6 @@ protected:
 
   /// \name Utilities for implementing subclasses.
   ///@{
-  static void addSystemFrameworkInclude(const llvm::opt::ArgList &DriverArgs,
-                                        llvm::opt::ArgStringList &CC1Args,
-                                        const Twine &Path);
   static void addSystemInclude(const llvm::opt::ArgList &DriverArgs,
                                llvm::opt::ArgStringList &CC1Args,
                                const Twine &Path);
@@ -238,9 +235,6 @@ protected:
       addExternCSystemIncludeIfExists(const llvm::opt::ArgList &DriverArgs,
                                       llvm::opt::ArgStringList &CC1Args,
                                       const Twine &Path);
-  static void addSystemFrameworkIncludes(const llvm::opt::ArgList &DriverArgs,
-                                         llvm::opt::ArgStringList &CC1Args,
-                                         ArrayRef<StringRef> Paths);
   static void addSystemIncludes(const llvm::opt::ArgList &DriverArgs,
                                 llvm::opt::ArgStringList &CC1Args,
                                 ArrayRef<StringRef> Paths);
diff --git a/clang/lib/Driver/Job.cpp b/clang/lib/Driver/Job.cpp
index 880e9e396..9573e5444 100644
--- a/clang/lib/Driver/Job.cpp
+++ b/clang/lib/Driver/Job.cpp
@@ -67,15 +67,14 @@ static bool skipArgs(const char *Flag, bool HaveCrashVFS, int &SkipNum,
     return true;
 
   // Some include flags shouldn't be skipped if we have a crash VFS
-  IsInclude =
-      llvm::StringSwitch<bool>(Flag)
-          .Cases("-include", "-header-include-file", true)
-          .Cases("-idirafter", "-internal-isystem", "-iwithprefix", true)
-          .Cases("-internal-externc-isystem", "-iprefix", true)
-          .Cases("-iwithprefixbefore", "-isystem", "-iquote", true)
-          .Cases("-isysroot", "-I", "-F", "-resource-dir", true)
-          .Cases("-internal-iframework", "-iframework", "-include-pch", true)
-          .Default(false);
+  IsInclude = llvm::StringSwitch<bool>(Flag)
+    .Cases("-include", "-header-include-file", true)
+    .Cases("-idirafter", "-internal-isystem", "-iwithprefix", true)
+    .Cases("-internal-externc-isystem", "-iprefix", true)
+    .Cases("-iwithprefixbefore", "-isystem", "-iquote", true)
+    .Cases("-isysroot", "-I", "-F", "-resource-dir", true)
+    .Cases("-iframework", "-include-pch", true)
+    .Default(false);
   if (IsInclude)
     return !HaveCrashVFS;
 
diff --git a/clang/lib/Driver/ToolChain.cpp b/clang/lib/Driver/ToolChain.cpp
index 07a3ae925..f1f6c3df8 100644
--- a/clang/lib/Driver/ToolChain.cpp
+++ b/clang/lib/Driver/ToolChain.cpp
@@ -1383,17 +1383,10 @@ ToolChain::CXXStdlibType ToolChain::GetCXXStdlibType(const ArgList &Args) const{
   return *cxxStdlibType;
 }
 
-/// Utility function to add a system framework directory to CC1 arguments.
-void ToolChain::addSystemFrameworkInclude(const llvm::opt::ArgList &DriverArgs,
-                                          llvm::opt::ArgStringList &CC1Args,
-                                          const Twine &Path) {
-  CC1Args.push_back("-internal-iframework");
-  CC1Args.push_back(DriverArgs.MakeArgString(Path));
-}
-
 /// Utility function to add a system include directory to CC1 arguments.
-void ToolChain::addSystemInclude(const ArgList &DriverArgs,
-                                 ArgStringList &CC1Args, const Twine &Path) {
+/*static*/ void ToolChain::addSystemInclude(const ArgList &DriverArgs,
+                                            ArgStringList &CC1Args,
+                                            const Twine &Path) {
   CC1Args.push_back("-internal-isystem");
   CC1Args.push_back(DriverArgs.MakeArgString(Path));
 }
@@ -1406,9 +1399,9 @@ void ToolChain::addSystemInclude(const ArgList &DriverArgs,
 /// "C" semantics. These semantics are *ignored* by and large today, but its
 /// important to preserve the preprocessor changes resulting from the
 /// classification.
-void ToolChain::addExternCSystemInclude(const ArgList &DriverArgs,
-                                        ArgStringList &CC1Args,
-                                        const Twine &Path) {
+/*static*/ void ToolChain::addExternCSystemInclude(const ArgList &DriverArgs,
+                                                   ArgStringList &CC1Args,
+                                                   const Twine &Path) {
   CC1Args.push_back("-internal-externc-isystem");
   CC1Args.push_back(DriverArgs.MakeArgString(Path));
 }
@@ -1420,28 +1413,19 @@ void ToolChain::addExternCSystemIncludeIfExists(const ArgList &DriverArgs,
     addExternCSystemInclude(DriverArgs, CC1Args, Path);
 }
 
-/// Utility function to add a list of system framework directories to CC1.
-void ToolChain::addSystemFrameworkIncludes(const ArgList &DriverArgs,
-                                           ArgStringList &CC1Args,
-                                           ArrayRef<StringRef> Paths) {
-  for (const auto &Path : Paths) {
-    CC1Args.push_back("-internal-iframework");
-    CC1Args.push_back(DriverArgs.MakeArgString(Path));
-  }
-}
-
 /// Utility function to add a list of system include directories to CC1.
-void ToolChain::addSystemIncludes(const ArgList &DriverArgs,
-                                  ArgStringList &CC1Args,
-                                  ArrayRef<StringRef> Paths) {
+/*static*/ void ToolChain::addSystemIncludes(const ArgList &DriverArgs,
+                                             ArgStringList &CC1Args,
+                                             ArrayRef<StringRef> Paths) {
   for (const auto &Path : Paths) {
     CC1Args.push_back("-internal-isystem");
     CC1Args.push_back(DriverArgs.MakeArgString(Path));
   }
 }
 
-std::string ToolChain::concat(StringRef Path, const Twine &A, const Twine &B,
-                              const Twine &C, const Twine &D) {
+/*static*/ std::string ToolChain::concat(StringRef Path, const Twine &A,
+                                         const Twine &B, const Twine &C,
+                                         const Twine &D) {
   SmallString<128> Result(Path);
   llvm::sys::path::append(Result, llvm::sys::path::Style::posix, A, B, C, D);
   return std::string(Result);
diff --git a/clang/lib/Driver/ToolChains/Darwin.cpp b/clang/lib/Driver/ToolChains/Darwin.cpp
index 234683f2f..8d469f7d1 100644
--- a/clang/lib/Driver/ToolChains/Darwin.cpp
+++ b/clang/lib/Driver/ToolChains/Darwin.cpp
@@ -2676,26 +2676,6 @@ void AppleMachO::AddClangSystemIncludeArgs(
   }
 }
 
-void DarwinClang::AddClangSystemIncludeArgs(
-    const llvm::opt::ArgList &DriverArgs,
-    llvm::opt::ArgStringList &CC1Args) const {
-  AppleMachO::AddClangSystemIncludeArgs(DriverArgs, CC1Args);
-
-  if (DriverArgs.hasArg(options::OPT_nostdinc, options::OPT_nostdlibinc))
-    return;
-
-  llvm::SmallString<128> Sysroot = GetEffectiveSysroot(DriverArgs);
-
-  // Add <sysroot>/System/Library/Frameworks
-  // Add <sysroot>/System/Library/SubFrameworks
-  // Add <sysroot>/Library/Frameworks
-  SmallString<128> P1(Sysroot), P2(Sysroot), P3(Sysroot);
-  llvm::sys::path::append(P1, "System", "Library", "Frameworks");
-  llvm::sys::path::append(P2, "System", "Library", "SubFrameworks");
-  llvm::sys::path::append(P3, "Library", "Frameworks");
-  addSystemFrameworkIncludes(DriverArgs, CC1Args, {P1, P2, P3});
-}
-
 bool DarwinClang::AddGnuCPlusPlusIncludePaths(const llvm::opt::ArgList &DriverArgs,
                                               llvm::opt::ArgStringList &CC1Args,
                                               llvm::SmallString<128> Base,
diff --git a/clang/lib/Driver/ToolChains/Darwin.h b/clang/lib/Driver/ToolChains/Darwin.h
index b38bfe6d1..76523d636 100644
--- a/clang/lib/Driver/ToolChains/Darwin.h
+++ b/clang/lib/Driver/ToolChains/Darwin.h
@@ -647,10 +647,6 @@ public:
   /// @name Apple ToolChain Implementation
   /// {
 
-  void
-  AddClangSystemIncludeArgs(const llvm::opt::ArgList &DriverArgs,
-                            llvm::opt::ArgStringList &CC1Args) const override;
-
   RuntimeLibType GetRuntimeLibType(const llvm::opt::ArgList &Args) const override;
 
   void AddLinkRuntimeLibArgs(const llvm::opt::ArgList &Args,
diff --git a/clang/lib/Frontend/CompilerInvocation.cpp b/clang/lib/Frontend/CompilerInvocation.cpp
index 08f3b7a7f..3744346e8 100644
--- a/clang/lib/Frontend/CompilerInvocation.cpp
+++ b/clang/lib/Frontend/CompilerInvocation.cpp
@@ -3405,8 +3405,6 @@ static void GenerateHeaderSearchArgs(const HeaderSearchOptions &Opts,
                            : OPT_internal_externc_isystem;
     GenerateArg(Consumer, Opt, It->Path);
   }
-  for (; It < End && Matches(*It, {frontend::System}, true, true); ++It)
-    GenerateArg(Consumer, OPT_internal_iframework, It->Path);
 
   assert(It == End && "Unhandled HeaderSearchOption::Entry.");
 
@@ -3539,8 +3537,6 @@ static bool ParseHeaderSearchArgs(HeaderSearchOptions &Opts, ArgList &Args,
       Group = frontend::ExternCSystem;
     Opts.AddPath(A->getValue(), Group, false, true);
   }
-  for (const auto *A : Args.filtered(OPT_internal_iframework))
-    Opts.AddPath(A->getValue(), frontend::System, true, true);
 
   // Add the path prefixes which are implicitly treated as being system headers.
   for (const auto *A :
diff --git a/clang/lib/Lex/InitHeaderSearch.cpp b/clang/lib/Lex/InitHeaderSearch.cpp
index 3e22b4001..bfe3da86b 100644
--- a/clang/lib/Lex/InitHeaderSearch.cpp
+++ b/clang/lib/Lex/InitHeaderSearch.cpp
@@ -244,9 +244,6 @@ bool InitHeaderSearch::ShouldAddDefaultIncludePaths(
     break;
   }
 
-  if (triple.isOSDarwin())
-    return false;
-
   return true; // Everything else uses AddDefaultIncludePaths().
 }
 
@@ -261,6 +258,22 @@ void InitHeaderSearch::AddDefaultIncludePaths(
   if (!ShouldAddDefaultIncludePaths(triple))
     return;
 
+  // NOTE: some additional header search logic is handled in the driver for
+  // Darwin.
+  if (triple.isOSDarwin()) {
+    if (HSOpts.UseStandardSystemIncludes) {
+      // Add the default framework include paths on Darwin.
+      if (triple.isDriverKit()) {
+        AddPath("/System/DriverKit/System/Library/Frameworks", System, true);
+      } else {
+        AddPath("/System/Library/Frameworks", System, true);
+        AddPath("/System/Library/SubFrameworks", System, true);
+        AddPath("/Library/Frameworks", System, true);
+      }
+    }
+    return;
+  }
+
   if (Lang.CPlusPlus && !Lang.AsmPreprocessor &&
       HSOpts.UseStandardCXXIncludes && HSOpts.UseStandardSystemIncludes) {
     if (HSOpts.UseLibcxx) {
diff --git a/clang/test/Driver/Inputs/DriverKit19.0.sdk/System/DriverKit/System/Library/SubFrameworks/.keep b/clang/test/Driver/Inputs/DriverKit19.0.sdk/System/DriverKit/System/Library/SubFrameworks/.keep
deleted file mode 100644
index e69de29bb..000000000
diff --git a/clang/test/Driver/Inputs/MacOSX15.1.sdk/Library/Frameworks/.keep b/clang/test/Driver/Inputs/MacOSX15.1.sdk/Library/Frameworks/.keep
deleted file mode 100644
index e69de29bb..000000000
diff --git a/clang/test/Driver/Inputs/MacOSX15.1.sdk/System/Library/Frameworks/.keep b/clang/test/Driver/Inputs/MacOSX15.1.sdk/System/Library/Frameworks/.keep
deleted file mode 100644
index e69de29bb..000000000
diff --git a/clang/test/Driver/Inputs/MacOSX15.1.sdk/System/Library/SubFrameworks/.keep b/clang/test/Driver/Inputs/MacOSX15.1.sdk/System/Library/SubFrameworks/.keep
deleted file mode 100644
index e69de29bb..000000000
diff --git a/clang/test/Driver/darwin-framework-search-paths.c b/clang/test/Driver/darwin-framework-search-paths.c
deleted file mode 100644
index 1cb4dc420..000000000
--- a/clang/test/Driver/darwin-framework-search-paths.c
+++ /dev/null
@@ -1,23 +0,0 @@
-// UNSUPPORTED: system-windows
-//   Windows is unsupported because we use the Unix path separator `/` in the test.
-
-// RUN: %clang %s -target arm64-apple-macosx15.1 -isysroot %S/Inputs/MacOSX15.1.sdk -c %s -### 2>&1 \
-// RUN: | FileCheck -DSDKROOT=%S/Inputs/MacOSX15.1.sdk %s
-//
-// CHECK: "-cc1"
-// CHECK: "-resource-dir" "[[RESOURCE_DIR:[^"]*]]"
-// CHECK-SAME: "-internal-iframework" "[[SDKROOT]]/System/Library/Frameworks"
-// CHECK-SAME: "-internal-iframework" "[[SDKROOT]]/System/Library/SubFrameworks"
-// CHECK-SAME: "-internal-iframework" "[[SDKROOT]]/Library/Frameworks"
-
-// Verify that -nostdlibinc and -nostdinc removes the default search paths.
-//
-// RUN: %clang %s -target arm64-apple-macosx15.1 -isysroot %S/Inputs/MacOSX15.1.sdk -nostdinc -c %s -### 2>&1 \
-// RUN: | FileCheck --check-prefix=CHECK-NOSTD -DSDKROOT=%S/Inputs/MacOSX15.1.sdk %s
-//
-// RUN: %clang %s -target arm64-apple-macosx15.1 -isysroot %S/Inputs/MacOSX15.1.sdk -nostdlibinc -c %s -### 2>&1 \
-// RUN: | FileCheck --check-prefix=CHECK-NOSTD -DSDKROOT=%S/Inputs/MacOSX15.1.sdk %s
-//
-// CHECK-NOSTD: "-cc1"
-// CHECK-NOSTD: "-resource-dir" "[[RESOURCE_DIR:[^"]*]]"
-// CHECK-NOSTD-NOT: "-internal-iframework"
diff --git a/clang/test/Driver/driverkit-path.c b/clang/test/Driver/driverkit-path.c
index 32f0a6721..3caae382d 100644
--- a/clang/test/Driver/driverkit-path.c
+++ b/clang/test/Driver/driverkit-path.c
@@ -21,15 +21,13 @@ int main() { return 0; }
 // LD64-NEW: "-isysroot" "[[PATH:[^"]*]]Inputs/DriverKit19.0.sdk"
 // LD64-NEW-NOT: "-L[[PATH]]Inputs/DriverKit19.0.sdk/System/DriverKit/usr/lib"
 // LD64-NEW-NOT: "-F[[PATH]]Inputs/DriverKit19.0.sdk/System/DriverKit/System/Library/Frameworks"
-// LD64-NEW-NOT: "-F[[PATH]]Inputs/DriverKit19.0.sdk/System/DriverKit/System/Library/SubFrameworks"
 
 
-// RUN: %clang %s -target x86_64-apple-driverkit19.0 -isysroot %S/Inputs/DriverKit19.0.sdk -x c++ -### 2>&1 \
-// RUN: | FileCheck %s -DSDKROOT=%S/Inputs/DriverKit19.0.sdk --check-prefix=INC
+// RUN: %clang %s -target x86_64-apple-driverkit19.0 -isysroot %S/Inputs/DriverKit19.0.sdk -E -v -x c++ 2>&1 | FileCheck %s --check-prefix=INC
 //
-// INC: "-isysroot" "[[SDKROOT]]"
-// INC: "-internal-isystem" "[[SDKROOT]]/System/DriverKit/usr/local/include"
-// INC: "-internal-isystem" "{{.+}}/lib{{(64)?}}/clang/{{[^/ ]+}}/include"
-// INC: "-internal-externc-isystem" "[[SDKROOT]]/System/DriverKit/usr/include"
-// INC: "-internal-iframework" "[[SDKROOT]]/System/DriverKit/System/Library/Frameworks"
-// INC: "-internal-iframework" "[[SDKROOT]]/System/DriverKit/System/Library/SubFrameworks"
+// INC:       -isysroot [[PATH:[^ ]*/Inputs/DriverKit19.0.sdk]]
+// INC-LABEL: #include <...> search starts here:
+// INC:       [[PATH]]/System/DriverKit/usr/local/include
+// INC:       /lib{{(64)?}}/clang/{{[^/ ]+}}/include
+// INC:       [[PATH]]/System/DriverKit/usr/include
+// INC:       [[PATH]]/System/DriverKit/System/Library/Frameworks (framework directory)
